endpoints/openrtb2/amp_auction.go

```diff
@@ -0,0 +1,230 @@
+// of the request, and the return value, using the OpenRTB machinery to handle everything inbetween.
+func NewAmpEndpoint(ex exchange.Exchange, validator openrtb_ext.BidderParamValidator, requestsById stored_requests.Fetcher, cfg *config.Configuration, met *pbsmetrics.Metrics) (httprouter.Handle, error) {
+	if ex == nil || validator == nil || requestsById == nil || cfg == nil || met == nil {
+		return nil, errors.New("NewEndpoint requires non-nil arguments.")
+	}
+
+	return httprouter.Handle((&endpointDeps{ex, validator, requestsById, cfg, met}).AmpAuction), nil
+}
+
+func (deps *endpointDeps) AmpAuction(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	// Prebid Server interprets request.tmax to be the maximum amount of time that a caller is willing
+	// to wait for bids. However, tmax may be defined in the Stored Request data.
+	//
+	// If so, then the trip to the backend might use a significant amount of this time.
+	// We can respect timeouts more accurately if we note the *real* start time, and use it
+	// to compute the auction timeout.
+	start := time.Now()
+
+	deps.metrics.RequestMeter.Mark(1)
+	// Set this as an AMP request in Metrics. Should it also count as an OpenRTB request?
+	deps.metrics.AmpRequestMeter.Mark(1)
+
+	req, errL := deps.parseAmpRequest(r)
+	isSafari := false
+	if ua := user_agent.New(r.Header.Get("User-Agent")); ua != nil {
+		name, _ := ua.Browser()
+		if name == "Safari" {
+			isSafari = true
+			deps.metrics.SafariRequestMeter.Mark(1)
+		}
+	}
+
+	if len(errL) > 0 {
+		w.WriteHeader(http.StatusBadRequest)
+		for _, err := range errL {
+			w.Write([]byte(fmt.Sprintf("Invalid request format: %s\n", err.Error())))
+		}
+		deps.metrics.ErrorMeter.Mark(1)
+		return
+	}
+
+	ctx := context.Background()
+	cancel := func() {}
+	if req.TMax > 0 {
+		ctx, cancel = context.WithDeadline(ctx, start.Add(time.Duration(req.TMax)*time.Millisecond))
+	} else {
+		ctx, cancel = context.WithDeadline(ctx, start.Add(time.Duration(5000)*time.Millisecond))
+	}
+	defer cancel()
+
+	usersyncs := pbs.ParsePBSCookieFromRequest(r, &(deps.cfg.HostCookie.OptOutCookie))
+	if req.App != nil {
+		deps.metrics.AppRequestMeter.Mark(1)
+	} else if usersyncs.LiveSyncCount() == 0 {
+		deps.metrics.NoCookieMeter.Mark(1)
+		if isSafari {
+			deps.metrics.SafariNoCookieMeter.Mark(1)
+		}
+	}
+
+	response, err := deps.ex.HoldAuction(ctx, req, usersyncs)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "Critical error while running the auction: %v", err)
+		glog.Errorf("/openrtb2/amp Critical error: %v", err)
+		return
+	}
+
+	// Need to extract the targeting parameters from the response, as those are all that
+	// go in the AMP response
+	targets := map[string]string{}
+	for _, seatBids := range response.SeatBid {
+		for _, bid := range seatBids.Bid {
+			if strings.Contains(string(bid.Ext), "\"hb_cache_id") {
```

Comment:
The `bid.ext` is likely to be much bigger than `hb_cache_id`... so I think this should be `bytes.Contains()`.

`string(someByteArray)` probably does a full data copy, unless there's some compiler magic, because strings are immutable and byte arrays aren't.